### PR TITLE
Workspace Refactoring

### PR DESCRIPTION
**Pull request by ikeough**
### Purpose

This PR moves graph layout, presets, and node to code out to separate extensions classes. It moves the undo recorder to separate .cs files, but keeps it as a partial class of `WorkspaceModel`.

It also moves and renames the json serialization methods recently added by @QilongTang to core. These are now the static method `WorkspaceModel.FromJson(...)` and the extension method `myWorkspace.ToJson(...)`.

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files **n/a**
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any. **n/a**
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@QilongTang

### CODE REVIEW

**Comment by mjkkirschner** [comment]
@ikeough do you have access to the self service CI EC job?

File: src/DynamoCore/Graph/Workspaces/CustomNodeWorkspaceModel.cs [hunk 1]
```diff
@@ -235,7 +235,7 @@ public bool IsVisibleInDynamoLibrary
 
         private bool isVisibleInDynamoLibrary;
 
-        protected override void RequestRun()
+        internal override void RequestRun()
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Workspaces/CustomNodeWorkspaceModel.cs]: whats the need for this change?

Comment (ikeough) [inline src/DynamoCore/Graph/Workspaces/CustomNodeWorkspaceModel.cs]: It was upgraded to `internal` in the base class as well, as it's called from an extension method now.

File: src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs [hunk 2]
```diff
@@ -1,57 +1,24 @@
...
+﻿using Dynamo.Core;
 using Dynamo.Engine;
 using Dynamo.Graph.Nodes.NodeLoaders;
-using Dynamo.Graph.Workspaces;
-using Dynamo.Models;
 using Dynamo.Scheduler;
 using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Text.RegularExpressions;
```
Comment (QilongTang) [inline src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs]: Aha, apparently your sorting plugin decides to put system usings after Dynamo usings which is default for most plugins. Either way is acceptable among developers, but just hope that we specify that in the coding standard.

File: src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs [hunk 3]
```diff
@@ -1,57 +1,24 @@
...
-
-            return ws;
-        }
-
-        /// <summary>
         /// Save a Workspace to json.
         /// </summary>
         /// <returns>A string representing the serialized WorkspaceModel.</returns>
-        public static string SaveWorkspaceToJson(WorkspaceModel workspace, LibraryServices libraryServices, 
+        public static string ToJson(this WorkspaceModel workspace, LibraryServices libraryServices,
```
Comment (QilongTang) [inline src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs]: Neat

**Comment by QilongTang** [comment]
@ikeough LGTM. I appreciate the fact your are making `toJson` and `fromJson` part of workspaceModel in this PR as I was planning to log a tech_debt on this refactoring. I do have sessionTraceData serialization changes in my branch depending on this PR. So :shipit: and I shall re-work the changes based on the new signatures.

**Comment by mjkkirschner** [comment]
@ikeough I saw one failing test on your run last night on the CI, was that fixed?

**Comment by ikeough** [comment]
@mjkkirschner I haven't fixed that test yet. Will do so before merging.

**Comment by ikeough** [comment]
I am leaving the one failing test `DynamoCoreWpfTests.RecordedTestsDSEngine.Defect_MAGN_904` outstanding as it has been verified to predate this work.

**Pull request merged**